src/Core/Config/Util/ConfigFileManager.php

```diff
@@ -144,82 +149,153 @@ private function loadStaticConfig(string $name): array
+			 *
+			 * Any exclusive locking (LOCK_EX) would need to wait until all LOCK_SHs are unlocked
+			 */
 			$configStream = fopen($filename, 'r');
-			if (flock($configStream, LOCK_SH)) {
-				$content = fread($configStream, filesize($filename));
-				if (!$content) {
-					throw new ConfigFileException(sprintf('Couldn\'t read file %s', $filename));
+			try {
+				if (flock($configStream, LOCK_SH)) {
+					$content = fread($configStream, filesize($filename));
+					if (!$content) {
+						throw new ConfigFileException(sprintf('Couldn\'t read file %s', $filename));
+					}
 				}
+			} finally {
+				// unlock and close the stream for every circumstances
 				flock($configStream, LOCK_UN);
+				fclose($configStream);
 			}
 
-			fclose($configStream);
-
+			/**
+			 * Evaluate the fetched content
+			 *
+			 * @note
+			 * Because `eval()` directly evaluates PHP content, we need to "close" the expected PHP content again with
+			 * the prefixed "?>". Now we're in plain HTML again and can evaluate any PHP file :-)
+			 */
 			$dataArray = eval('?>' . $content);
 
-			if (!is_array($dataArray)) {
-				throw new ConfigFileException(sprintf('Error loading config file %s', $filename));
+			if (is_array($dataArray)) {
+				$configCache->load($dataArray, Cache::SOURCE_DATA);
 			}
-
-			$config->load($dataArray, Cache::SOURCE_DATA);
 		}
 	}
 
 	/**
-	 * Saves overridden config entries back into the data.config.phpR
+	 * Saves overridden config entries back into the data.config.php
 	 *
-	 * @param Cache $config The config cache
+	 * @param Cache $configCache The config cache
 	 *
 	 * @throws ConfigFileException In case the config file isn't writeable or the data is invalid
 	 */
-	public function saveData(Cache $config)
+	public function saveData(Cache $configCache)
 	{
-		$data = $config->getDataBySource(Cache::SOURCE_DATA);
-
-		$encodedData = ConfigFileTransformer::encode($data);
+		$filename = $this->configDir . '/' . self::CONFIG_DATA_FILE;
 
-		if (!$encodedData) {
-			throw new ConfigFileException('config source cannot get encoded');
+		if (file_exists($filename)) {
+			$fileExists = true;
+		} else {
+			$fileExists = false;
 		}
 
-		$configStream = fopen($this->configDir . '/' . self::CONFIG_DATA_FILE, 'w');
+		/**
+		 * Creates a read-write stream
+		 *
+		 * @see https://www.php.net/manual/en/function.fopen.php
+		 * @note Open the file for reading and writing. If the file does not exist, it is created.
+		 * If it exists, it is neither truncated (as opposed to 'w'), nor the call to this function fails
+		 * (as is the case with 'x'). The file pointer is positioned on the beginning of the file.
+		 *
+		 */
+		$configStream = fopen($filename, 'c+');
+
+		try {
+			// We do want an exclusive lock, so we wait until every LOCK_SH (config reading) is unlocked
+			if (flock($configStream, LOCK_EX)) {
+
+				/**
+				 * If the file exists, read the whole file again
```

Comment:
```suggestion
				 * If the file exists, we read the whole file again to avoid a race condition with concurrent threads that could have modified the file between the first config read of this thread and now
```